Activity Stream Phase 1 - Layout

farhanpatel opened the pull request:
I'd love for some feedback. I'll comment on some design decisions I made in #2039

farhanpatel commented on Client/Frontend/Home/ActivityStreamSupplementary.swift:
```diff
@@ -0,0 +1,310 @@
+                return
+            }
+
+            img.getColors(CGSize(width: 50, height:50)) { colors in
```
I'm using UIImageColors here. It extracts dominant colors form images. Works pretty great.

farhanpatel commented on Client/Frontend/Home/ActivityStreamSupplementary.swift:
```diff
@@ -0,0 +1,310 @@
+        }
+    }
+
+    func configureWithTopSiteItem(site: TopSiteItem) {
```
I've moved configuration of the cell into its own function. What do you think of this approach compared to customizing the cell in cellForRowAtIndexPath?

farhanpatel commented on Client/Frontend/Home/ActivityStreamSupplementary.swift:
```diff
@@ -0,0 +1,310 @@
+    private var pageControl: UIPageControl!
+    private var headerView: ASHeaderView!
+
+    var headerTitle: String = "" {
```
Instead of letting external classes mess with headerView. I created a headerTitle that will update the headerView's title

farhanpatel commented on Client/Frontend/Home/ActivityStreamSupplementary.swift:
```diff
@@ -0,0 +1,310 @@
+        collectionView.delegate = delegate
+        collectionView.dataSource = delegate
+        delegate.numberOfPagesChangedHandler = numberOfPagesChanged
+        delegate.pageChangedHandler = currentPageChanged
```
I created function handlers to help observe changes to the collectionview which can update pagecontrol. This way I dont expose pageControl outside of its class. I wish there were properties I could observe directly on collectionview without having to make the cell a delegate. But couldnt find any. Thoughts on this approach compared to just exposing pagecontrol?

farhanpatel commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+        self.profile = profile
+        super.init(nibName: nil, bundle: nil)
+
+        self.profile.history.setTopSitesCacheSize(Int32(ASPanelUX.topSitesCacheSize))
```
Instead of the old topsites where the controller did view calculations to figure out the cache size. I just set a large enough default and then only display a certain amount. Keeps the VC cleaner.

farhanpatel commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+        }
+    }
+
+    override func traitCollectionDidChange(previousTraitCollection: UITraitCollection?) {
```
By keeping track of the traits I can make sure that the right amount of topsites appear for all view sizes.

sleroux replied:
Make sure to always call `super` whenever overriding view controller methods.

farhanpatel commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+}
+
+// MARK: -  Section management
+extension ActivityStreamPanel {
```
I saw this enum configuration in an Apple sample project. Works pretty well for static table settings.

farhanpatel commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+extension ActivityStreamPanel {
+
+    func tableView(tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
+        return Section(section).headerHeight
```
So simple 👍

sleroux replied:
Such clean-ness 👌

farhanpatel commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+    func configureTopSitesCell(cell: UITableViewCell, forIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let topSiteCell = cell as! ASHorizontalScrollCell
+        // The topSiteCell needs a refrence to the tableview because it needs to alert the tableView to relayout once topsites finishes loading.
+        topSiteCell.parentTableView = self.tableView
```
This is one of my least favorite hacks. there was no way I could think of to relayout the tableview without adding this reference 😢

farhanpatel commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+        return simpleHighlightCell
+    }
+
+    func showSiteWithURL(url: NSURL) {
```
Instead of passing a homepanel refrence to topsites. I pass closure like this.

farhanpatel commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+}
+
+// MARK: - Data Management
+extension ActivityStreamPanel {
```
this stuff is copy pasted from old topsites.

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+
+class ActivityStreamPanel: UIViewController, UICollectionViewDelegate {
+    weak var homePanelDelegate: HomePanelDelegate? = nil
+    let profile: Profile
```
Nit: Probably better if this is private. We shouldn't need to access this outside of the panel.

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+        return tableView
+    }()
+
+    var topSiteHandler = ASHorizontalScrollSource()
```
Nit: Private

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+    static let historySize = 10
+}
+
+class ActivityStreamPanel: UIViewController, UICollectionViewDelegate {
```
Nit: Don't need `UICollectionViewDelegate` anymore

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+extension ActivityStreamPanel {
+    enum Section: Int {
+        case topSites
+        case history
```
Nit: Using lowercased enum cases is a Swift 3 style. Not sure if we want to move to this yet or not. Maybe keep it capital for consistency?

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+        var cellIdentifier: String {
+            switch self {
+            case .topSites: return "TopSite"
+            default: return "Cell"
```
Nit: For theses switch statements, prefer exhaustive cases instead of relying on default - even if we're returning nil. This way if we add anything to the enum we can lean on the compiler to yell at us instead of having behaviour we might not want.

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+
+        init(_ section: Int) {
+            self.init(rawValue: section)!
+        }
```
Might be safer to use a fail-able initializer here instead of forcing the conversion.

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+}
+
+// MARK: -  Header Views
+extension ActivityStreamPanel {
```
These are `UITableViewDelegate` methods but the extension doesn't make that clear and the protocol is used in the extension below. Any chance we could move the delegate methods here from the extension below and separate the delegate from the data source?

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        switch Section(section) {
+            case .topSites:
+                if !topSiteHandler.content.isEmpty {
```
Nit: Would read a bit better if you didn't use the negative:

```swift

reutrn topSiteHandler.content. isEmpty ? 0 : 1
``

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+        switch Section(indexPath.section) {
+            case .topSites:
+                return configureTopSitesCell(cell, forIndexPath: indexPath)
+            default:
```
Nit: Opt for exhaustive cases instead of default

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+    }
+
+    func showSiteWithURL(url: NSURL) {
+        let visitType = VisitType.Bookmark
```
How come this is always a bookmark?

farhanpatel replied:
I was just following how topsites does it. But looking at the options on VisitType. Bookmark works the best. 
https://github.com/mozilla/firefox-ios/blob/0642d2c967c85812b5918a5d7598a707cad70cdd/Client/Frontend/Home/TopSitesPanel.swift#L354

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+        case .history:
+            let site = self.history[indexPath.row]
+            showSiteWithURL(NSURL(string:site.url)!)
+        default:
```
Nit: default -> .topSites

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+            self.refreshTopSites(ASPanelUX.topSitesCacheSize)
+        default:
+            // no need to do anything at all
+            print("Dont have this notification type")
```
Not sure if this is temporary or not but when logging use an instance of XCLogger from Logger.swift instead of print.

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+    }()
+
+    var title: String = "" {
+        willSet(newTitle) {
```
Nit: Not sure if it makes a difference but logically I think of this as 'after setting the title, also set the label'  which would be using `didSet` instead `willSet`.

farhanpatel replied:
for some reason. didSet doesnt update the label correctly. dunno why :(

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+            }
+            return succeed()
+        }
+    }
```
Looks like this was borrowed from the existing top sites. It's a bit confusing the way we interweave between background/main queue processing when updating sites. I did a minor rewrite of this that might make it more clear. It keeps things that can be done on the background in the background while grouping all the potentially main thread related things in the uponQueue block.

```swift

    private func reloadRecentHistory() {
        self.profile.history.getSitesByLastVisit(ASPanelUX.historySize).uponQueue(dispatch_get_main_queue()) { result in
            self.history = result.successValue?.asArray() ?? self.history
            self.tableView.reloadData()
        }
    }

    private func reloadTopSites() {
        invalidateTopSites().uponQueue(dispatch_get_main_queue()) { result in
            let sites = result.successValue ?? []
            self.topSites = sites
            self.topSiteHandler.content = self.topSites
            self.topSiteHandler.urlPressedHandler = { [unowned self] url in
                self.showSiteWithURL(url)
            }
            self.topSiteHandler.currentTraits = self.traitCollection
            self.tableView.reloadData()
        }
    }

    private func invalidateTopSites() -> Deferred<Maybe<[TopSiteItem]>> {
        let frecencyLimit = ASPanelUX.topSitesCacheSize
        return self.profile.history.areTopSitesDirty(withLimit: frecencyLimit) >>== { dirty in
            if dirty || self.topSites.isEmpty {
                return self.profile.history.getTopSitesWithLimit(frecencyLimit) >>== { topSites in
                    return deferMaybe(topSites.flatMap(self.siteToItem))
                }
            }
            return deferMaybe(self.topSites)
        }
    }

    private func siteToItem(site: Site?) -> TopSiteItem? {
        guard let site = site else {
            return nil
        }

        guard let faviconURL = site.icon?.url else {
            return TopSiteItem(urlTitle: site.tileURL.extractDomainName(), faviconURL: nil, siteURL: site.tileURL)
        }

        return TopSiteItem(urlTitle: site.tileURL.extractDomainName(), faviconURL: NSURL(string: faviconURL)!, siteURL: site.tileURL)
    }
```

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,337 @@
+        return cell
+    }
+
+    func configureHistoryItemCell(cell: UITableViewCell, forIndexPath indexPath: NSIndexPath) -> UITableViewCell {
```
Nit: Just noticed you have a `configureSampleHighglithCell` method on the `SimpleHighlightCell`. You can probably get rid of this method and just do the following in the switch. I'd also renamed the configure method to be more specific and less redundant like `configureWithSite`.

```swift
switch Section(indexPath.section) {
case .topSites:
    let site = history[indexPath.row]
    (cell as? SimpleHighlightCell)?.configureCellWithSite(site)
case .history:
    ...
}
...
return cell
```

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,392 @@
+            make.height.equalTo(newHeight).priorityMedium()
+        }
+        self.layoutSubviews()
+        parentTableView?.reloadData()
```
Ah I see now about your comment about the hack in `ActivityStreamPanel` where you assign the parentTableView to the cell. If I understand correctly (let me know if I'm totally off base),  but this is similar in how we handle layout for cells on the old top sites where the cells scale which in turn inform the number of rows/columns that we show. This in turn dictates how high the `ASHoriziontalScrollView` should be thus the need to inform our tableView. I know at least in the old top sites logic this was hard to reason about and deal with.

What if we inversed the relationship? Instead of having this chain of cell -> ASHorizontalScrollCell -> UITableView, what if the table view tells the row, fit in X amount space. The scroll view and cells would then adjust accordingly. I think by having the table view dictate the terms of sizing, it will be easier to update the layout based on view controller lifecycle events and trait collections since the table view and view controller are next together in the hierarchy. Also would give us better control over exactly how we want to render top sites for various phones/layouts/sizes.

farhanpatel replied:
Yea you are exactly right. But short of hard coding in default heights for each different device type we are stuck in the same boat as the previous TopSites. 

We need to guess a height in the VC. If we want a well educated guess we need to use view properties defined at lower level such as the size of each item / the spacing and the number of items per row. But at this point we are basically doing the view calculations the layout performs. That is what we did in the old Topsites and it got hairy having all the view layout code in the VC. I dont know how else we could get the tableview to dictate the height short of hardcoding in specific heights for devices.

sleroux replied:
> But short of hard coding in default heights for each different device type we are stuck in the same boat as the previous TopSites. 

How about we hard code the height for the Top Sites section? We could have set of various heights for each trait collection (compact/regular). It would also let us be exact with the measurements from the designs without having to auto-infer anything. If hard coding the heights for the section makes everything else align nicely and fall into place architecturally it might make sense and simplify the code greatly.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,392 @@
+        pageControl.progress = currentPage
+    }
+
+    func setDelegate(delegate: ASHorizontalScrollSource) {
```
Nit: Seems weird to have a `setDelegate` method - feels Objective-Cish. Why not make it a property?

```swift
weak var delegate: ASHorizontalScrollSource {
   didSet {
       ...
   }
}
```

Also I'm not sure about the side effect of reloading the collection view here.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,392 @@
+    private var cellCount = 0
+    private var boundsSize = CGSizeZero
+    private var insets = UIEdgeInsetsZero
+    private let minimumInsets: CGFloat = 20
```
Nit: Use swift versions of CGSizeZero, UIEdgeInsetsZero

```swift
    CGSize.zero, UIEdgeInsets.zero
```

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,392 @@
+    var pageChangedHandler: ((CGFloat) -> Void)?
+
+    // The current traits that define the parent ViewController. Used to determine how many rows/columns should be created.
+    var currentTraits: UITraitCollection!
```
Try to avoid explicitly unwrapped variables. I think there is a way we can get away with this either not being optional or let it be optional and handle that throughout the code.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,392 @@
+
+/*
+ This Delegate/DataSource is used to manage the ASHorizontalScrollCell's UICollectionView. 
+ This is left generic enough for it to be re used for other parts of Activity Stream.
```
If you're interested in making this generic, you could consider throwing in a generic parameter for the content property. You might need to create a base protocol for 'things that a cell can configure with' to constrain the generic type into something more meaningful:

```swift
 +class ASHorizontalScrollSource<T: ViewModel>: NSObject, UICollectionViewDelegate, UICollectionViewDataSource, ASHorizontalScrollDelegate { 
    var content: [T] = []
    ...
    cell.configureWithViewModel(model)
}
```

But! If we don't need this to be generic, then we probably shouldn't bother until we need it.

farhanpatel replied:
We dont need this to be generic yet. But later (based on the mocks) there are places where I think we can reuse ASHorizontalScrollCell. I'll make it generic then.

farhanpatel commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,354 @@
+            }
+        }
+
+        func cellHeight(traits: UITraitCollection) -> CGFloat {
```
Here is the new way CellHeights are calculated for topSites. It work better and I dont need to keep reloading tableview anymore 👍  Let me know what you think

sleroux replied:
Nice! This looks like it's much easier to work with instead of the loop we were doing before with reloading the table view.

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,351 @@
+        super.viewDidLoad()
+        reloadTopSites()
+
+        reloadRecentHistory()
```
I wonder if we should move this to `viewWillAppear` instead. `reloadTopSites` has some logic for not invalidating when not dirty. Would also make changes propagate easier when switching between panels.

sleroux replied:
Also, even though the `reload*` methods are async, I would feel more comfortable having the adding/setting up of the table view happen before we call reload since they require the table view to be there. Since `tableView` is also lazy, this means that initialization might happen while reloading instead of the `view.addSubview` call below.

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,351 @@
+        var cellIdentifier: String {
+            switch self {
+            case .TopSites: return "TopSite"
+            case .History: return "Cell"
```
Nit: Super minor but maybe use a more descriptive identifier like `HistoryCell`?

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,351 @@
+        return Section(indexPath.section).cellHeight(self.traitCollection)
+    }
+
+    func showSiteWithURL(url: NSURL) {
```
Nit: I would make this private to help identify it has a 'helper' method instead of something that is from the delegate.

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,351 @@
+        case NotificationProfileDidFinishSyncing:
+            // Only reload top sites if there the cache is dirty since the finish syncing
+            // notification is fired everytime the user re-enters the app from the background.
+            self.profile.history.areTopSitesDirty(withLimit: ASPanelUX.topSitesCacheSize) >>== { dirty in
```
We already have the `dirty` check within `invalidateTopSites`. Does it make sense to have this here?

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -0,0 +1,351 @@
+// MARK: - HomePanel Protocol
+extension ActivityStreamPanel: HomePanel {
+
+    func endEditing() {
```
`endEditing` is optional so we can just get rid of this extension.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+    }
+
+    override func prepareForReuse() {
+        contentView.backgroundColor = UIColor.lightGrayColor()
```
Make sure to call super on override UIKit methods.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+
+    func configureWithTopSiteItem(site: TopSiteItem) {
+        titleLabel.text = site.urlTitle
+        if let favURL = site.faviconURL {
```
Nit: Would read better as a `guard` since you want the favicon to normally be there but otherwise, just use the default favicon.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+        didSet {
+            collectionView.delegate = delegate
+            collectionView.dataSource = delegate
+            delegate!.pageChangedHandler = { [weak self] progress in
```
If we set the delegate to `nil` this will burn the house down 😟 Probably safer to use `?`

sleroux replied:
It's also declared as `weak` so we have no control over retaining this.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+    private let minimumInsets: CGFloat = 20
+    var numberOfPages = 0
+
+    override func prepareLayout() {
```
Super

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+            return horizontalItemsCount
+        }
+    }
+    func collectionViewSizeForRect(contentSize: CGSize) -> CGSize {
```
Nit: Since this method is only invoked from `collectionViewContentSize`, can we just remove this method and put the logic in that?

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+        numberOfPages = Int(ceil(Double(cellCount) / Double(itemsPerPage)))
+
+
+        insets = UIEdgeInsets(top: verticalInsets, left: horizontalInsets, bottom: verticalInsets, right: horizontalInsets)
```
Nit: Maybe just personal taste but how come there are double newlines throughout? Do we really need two lines of emptiness?

farhanpatel replied:
No thats just my bad. I've been deleting stuff and then not removing the space :(

farhanpatel replied:
Looks like there is a rule for this in swiftlint 😀  
https://github.com/realm/SwiftLint/issues/548 
We have about 115 violations. Not bad actually. I'll enable this sometime soon.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+    }
+
+    override func shouldInvalidateLayoutForBoundsChange(newBounds: CGRect) -> Bool {
+        if newBounds.width < 320 {
```
Nit: Simplify `return newBounds.width <= 320`

Also, not all iPhones are 320 in width. The 6 and 6S are larger [1]. Maybe drive the `shouldInvalidate` when the trait collection differs instead of a fixed width?

[1] https://www.paintcodeapp.com/news/iphone-6-screens-demystified

farhanpatel replied:
When the screen width is 320 I do want to invalidate layout. 
Sometimes the bounds of the collectionview becomes 100,75 or something very tiny. It usually happens when the orientation changes with the topsites not on screen. In those cases I dont want to invlidate the existing layout. I'll add a comment to explain what is going on here.

farhanpatel replied:
Okay I see what you mean. I can clean this up a bit more.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+
+    func computeLayoutAttributesForCellAtIndexPath(indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes {
+        let row = indexPath.row
+        let bounds = self.collectionView!.bounds
```
Try to avoid explicits. You can usually use a `guard` clause instead and return a default state or add additional logging if you're worried about it being nil.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+
+        let attr = UICollectionViewLayoutAttributes(forCellWithIndexPath: indexPath)
+
+        var frame = CGRectZero
```
Nit: Swift-ify `CGRect.zero`

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+    static let verticalItemsForTraitSizes = [UIUserInterfaceSizeClass.Compact : 1, UIUserInterfaceSizeClass.Regular : 2]
+    static let horizontalItemsForTraitSizes = [UIUserInterfaceSizeClass.Compact : 3, UIUserInterfaceSizeClass.Regular : 5]
+    static let maxNumberOfPages = 2
+    static let CellIdentifier = "TopSiteCell"
```
In the panel code you used `TopSite` as a cell identifier and here you used `TopSiteCell`. Are these supposed to the same? I think this is supposed to be the actual item cell right? In that case I would maybe rename it to `TopSiteItemCell` to clear it up.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+
+    var content: [TopSiteItem] = []
+    var urlPressedHandler: ((NSURL) -> Void)?
+    var pageChangedHandler: ((CGFloat) -> Void)?
```
Instead of defining closures here for handlers, why not put them into the delegate? It's already defined and would fit in nicely with existing UIKit methods.

farhanpatel replied:
Not quite sure I understand this. The handlers are here because ASHoriziontalSource is the data source for the topsites. I didnt want to pass the homePanelDelegate to it. 
If you mean why arent the clousures in the ASHorizontalScrollDelegate thats because it is the layout delegate and is only responsible for layout stuff. I'll rename it to ASHorizontalScrollLayoutDelegate

sleroux replied:
Sorry I think I was getting confused about how these handlers and delegates work. I took a closer look and I think I understand whats going on here. The handlers should be fine but I found a couple of things that were a bit confusing when reading through this.

1. It might be more clear to also define a `ASHorizontalScrollDataSource` along with the `ASHorizontalScrollDelegate`. Since you're following the UICollectionViewDS/Del convention, smooshing both of these under a 'delegate' is a bit of a misnomer.

2. I would also clarify what the delegate/data source do by separating them out into their own extensions off of `ASHorizontalScrollSource`. It might be a bit overkill considering that the `ASHorizontalScrollView` at the moment requires a `UICollectionView` so you can probably have the protocols extend from the `UICollectionView` ones.

3. Minor thing but I would rename `topSiteHandler` to something more reflective of how it's both a datasource and delegate for the top sites cell since you use the handler prefix for callback methods inside the `topSiteHandler`. Maybe `topSiteSource`?

I also tried to think of how the handlers would work as delegate methods but it's a bit strange because the `ActivityStreamPanel` needs the tap event while the cell itself needs the page change event. One option could be to define different delegates `ASHorizontalScrollInteractionDelegate` / `ASHorizontalScrollPageDelegate` but considering that we have one method for each one that's a bit crazy and pointless so stick with the handlers 👍

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+        Size classes define how many items to show per row/column.
+    */
+    func numberOfVerticalItems(estimatedItems: Int) -> Int {
+        if let traits = currentTraits {
```
Nit: I think I mentioned this above but a rule of thumb I generally use for when to use `if` vs `guard` is if you're conditioning on the value usually being there, opt for `guard` so the code you expect to usually run is not buried inside a code branch. I guess in general it just reduces the branching complexity of the code by flattening it out.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+    func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        let layout = collectionView.collectionViewLayout as! HorizontalFlowLayout
+        let maxItems = layout.maxVerticalItemsCount() * layout.maxHorizontalItemsCount() * ASTopSiteSourceUX.maxNumberOfPages
+        return min(maxItems, self.content.count)
```
This kind of feels weird. Why would `self.content.count` ever be higher than the max number of items? I feel like this was added for safety but shouldn't be needed if we're calculating our layout correctly.

farhanpatel replied:
Because we fetch more topsites than we actually need. The fetching ASPanel doesnt know exactly how many items to fetch.
Therefore when displaying we might have more items than we should display. In that case we return a smaller amount. I'll add a comment for this.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -0,0 +1,393 @@
+    func numberOfHorizontalItems(estimatedItems: Int) -> Int {
+        if let traits = currentTraits {
+            // An iPhone 5 in both landscape/portrait is considered compactWidth which means we need to let the layout determine how many items to show based on actual width.
+            return max(ASTopSiteSourceUX.horizontalItemsForTraitSizes[traits.horizontalSizeClass]!, estimatedItems)
```
Shouldn't normally be an issue but force unwrapping this might becoming trouble if Apple introduces some weird size class in the future. I think we can move things around to clean this up.

Since the traitCollection is the main input into determining a set of values, I would toss `ASTopSiteSourceUX` and create  a data structure to hold the values like this:

```swift
struct ASHorizontalScrollSourceAttributes {
    let verticalItemCount: Int
    let horizontalItemCount: Int
    let maxPageCount: Int
}
```

.. and include a static constructor that outputs what you need for a trait collection:

```swift
static func attributesForTraitCollection(traitCollection: UITraitCollection) -> ASHorizontalScrollSourceAttributes? {
    ...
}
```

Lastly, you can then be more declarative by create private local instances of the different attributes you'd want:

```swift
private let wideAttributes = ASHorizontalScrollSourceAttributes(
    verticalItemCount: 1,
    horizontalItemCount: 5,
    maxPageCount: 3
)
```

You can might be able to even expose the counts as computed properties that wrap an instance of the attributes struct instead of having them as methods in the delegate.

sleroux commented on Client/Frontend/Widgets/SimpleHighlightCell.swift:
```diff
@@ -0,0 +1,191 @@
+        titleLabel.font = DynamicFontHelper.defaultHelper.DeviceFontMediumBold
+        titleLabel.textColor = SimpleHighlightCellUX.LabelColor
+        titleLabel.textAlignment = .Left
+        titleLabel.numberOfLines = 3
```
What happens if we have a title longer than 3 lines?

farhanpatel replied:
It just truncates and adds the ... 
Based on the mocks.

sleroux commented on Client/Frontend/Widgets/SimpleHighlightCell.swift:
```diff
@@ -0,0 +1,191 @@
+        titleLabel.font = DynamicFontHelper.defaultHelper.DeviceFontDescriptionActivityStream
+        titleLabel.textColor = SimpleHighlightCellUX.DescriptionLabelColor
+        titleLabel.textAlignment = .Left
+        titleLabel.numberOfLines = 1
```
What happens if this is longer than 1 line?

bkmunar replied:
I'm assuming it just truncates the text - but they're just 1 - 2 words (e.g. Bookmarked, Open Tab, Visited, Reading List) so hopefully nothing would get chopped off(?) Would ping Bryan Bell for more info if that were to happen

sleroux commented on Client/Frontend/Widgets/SimpleHighlightCell.swift:
```diff
@@ -0,0 +1,191 @@
+
+    lazy var descriptionLabel: UILabel = {
+        let titleLabel = UILabel()
+        titleLabel.setContentHuggingPriority(1000, forAxis: UILayoutConstraintAxis.Vertical)
```
How come you need to set the hugging priority to required for this? Was there issues with the label squishing?

sleroux commented on Client/Frontend/Widgets/SimpleHighlightCell.swift:
```diff
@@ -0,0 +1,191 @@
+        siteImageView.layer.masksToBounds = true
+    }
+
+    func configureSimpleHighlightCell(site: Site) {
```
Nit: Since the cell is named `SimpleHighlightCell` have that in the name of the method is a bit redundant. How about `configureCell`?

sleroux commented on Client/Frontend/Widgets/SimpleHighlightCell.swift:
```diff
@@ -0,0 +1,191 @@
+            let url = icon.url
+            self.siteImageView.layer.borderWidth = 0
+            self.setImageWithURL(NSURL(string: url)!)
+        } else if let url = NSURL(string: site.url) {
```
Since we're didn't override `prepareForReuse`, if this condition fails, the previous image before the recycle will still be here. Let's override `prepareForReuse` to catch these cases.

sleroux commented on Client/Frontend/Widgets/SimpleHighlightCell.swift:
```diff
@@ -0,0 +1,191 @@
+        }
+        self.titleLabel.text = site.title.characters.count <= 1 ? site.url : site.title
+        configureCellStatus(site)
+        self.timeStamp.text = NSDate.fromMicrosecondTimestamp((site.latestVisit?.date)!).toRelativeTimeConciseString()
```
The explicit unwrap after a condition unwrap makes me nervous. I would `guard` this instead of `!`. The default state can be a blank string if it doesn't exist.

sleroux commented on ThirdParty/FilledPageControl.swift:
```diff
@@ -0,0 +1,154 @@
+
+import UIKit
+
+@IBDesignable public class FilledPageControl: UIView {
```
So there is a built in UIKit class for this named `UIPageControl` [1]. Feels weird to pull in more third party source just for a couple of dots.

[1] https://developer.apple.com/library/tvos/documentation/UIKit/Reference/UIPageControl_Class/

farhanpatel replied:
I cant customize the UIPagecontrol the way I want based on the mocks. The built in one only allows you to change the color. In the mocks you'll notice that the pagecontrol has a hollow center when not selected which is different than UIPageControl

sleroux commented on ThirdParty/UIImageColors.swift:
```diff
@@ -0,0 +1,256 @@
+//
+//  Created by Jathu Satkunarajah (@jathu) on 2015-06-11 - Toronto
+//  Original Cocoa version by Panic Inc. - Portland
+//
```
For any additional third party sources we add, make sure to also update our license page to give credit and display the author's license in Licenses.html

sleroux commented on Utils/TimeConstants.swift:
```diff
@@ -79,6 +79,43 @@ extension NSDate {
         return String(format: NSLocalizedString("just now", comment: "Relative time for a tab that was visited within the last few moments."))
     }
+
+    public func toRelativeTimeConciseString() -> String {
```
This seems almost exactly the same as the method above. Is there any way we can combine them? Maybe adding a concise flag we pass in to rejig the string we get back?

sleroux commented:
You mentioned you already found a couple of bugs but I'll just note the ones I came across here just in case.

* When the user scrolls the AS panel we should hide the keyboard to show all of the panel. I think the old top sites panel had a tap gesture for this.
* Since we're not hiding the keyboard when the user scrolls, when reaching the bottom of the panel, there are still items to be scrolled that are hidden behind the keyboard. We should inset the table view to handle the new viewport when the keyboard is up. We have used the `KeyboardHelper` class in the past to handle that but that's kind of janky. The best way of handling this is to actually convert the panel to a `UITableViewController` if possible. It will handle all of this for free!
* Missing highlight tap states for the TopSiteItems
* Probably noticed already but the background colors of the top site items keeps changing whenever I swipe between pages.

sleroux commented:
@farhanpatel @bkmunar Back to your guys! I would normally add a feedback+ flag but this PR is on 3 different bugs so I'll just say it here.

farhanpatel commented:
@sleroux Should be ready to go! 

I dont think the keyboard should be hidden when scrolling AS. Imagine these cases

Open new tab  -> Tap url bar -> (you decide to search through AS.)  ->you dont find what you are looking for. -> now you gotta tap the url again to type

Although I'll leave it to bbell to decide once he gets to play with it.

farhanpatel commented:
I cleaned up the naming a bit. Renamed `ASHorizontalScrollCellSource` to `ASHorizontalScrollCellManager` I didnt want to split it up into delegate/datasource because combined its only 60 lines and not that complicated. 

Also renamed `ASHorizontalDelegate` to `ASHorizontalLayoutDelegate` to signify that its really just a few helper methods used by the layout.

sleroux commented on Client/Frontend/Home/ActivityStreamPanel.swift:
```diff
@@ -69,10 +52,18 @@ class ActivityStreamPanel: UIViewController, HomePanel {
+        tableView.delegate = self
+        tableView.dataSource = self
+        tableView.rowHeight = UITableViewAutomaticDimension
+        tableView.separatorInset = UIEdgeInsetsZero
```
Nit: `UIEdgeInsets.zero` swift style

farhanpatel replied:
This only exists in ios10 https://developer.apple.com/reference/uikit/uiedgeinsets/1771772-zero

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -323,51 +323,67 @@ struct ASTopSiteSourceUX {
+class ASHorizontalScrollCellManager: NSObject, UICollectionViewDelegate, UICollectionViewDataSource, ASHorizontalLayoutDelegate {
+
+    var _content = [TopSiteItem]()
+    var content: [TopSiteItem] {
```
This seems misleading to me. When I assign items to `content` I don't expect some of the items to not make it into the list. I can see this being confusing for someone trying to assign to this property. I would try to make the side effect known either by moving the check outside this setter at the call site or creating a method instead of a computed property that expresses what happens.

farhanpatel replied:
I've removed this. It has weird issues with items going missing when switching orientations. Now I just always get 12 items which makes sure that there are enough items for each orientation.

sleroux commented on Client/Frontend/Home/ActivityStreamTopSitesCell.swift:
```diff
@@ -57,17 +57,30 @@ class TopSiteItemCell: UICollectionViewCell {
+        titleLabel.snp_remakeConstraints { make in
+            make.edges.equalTo(contentView).inset(UIEdgeInsets(top: heightInset, left: 0, bottom: 0, right: 0))
+        }
+        imageView.snp_remakeConstraints { make in
```
Not sure if this was in `layoutSubviews` before but you should avoid mixing autolayout code with `layoutSubviews` calls or things can get a bit messy/weird. If you're looking for a frequent way to update the constraints, how about `updateConstraints`?

sleroux commented on ThirdParty/UIImageColors.swift:
```diff
@@ -1,10 +1,31 @@
+//    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+//    SOFTWARE.
```
Actually I meant adding this to our `Licenses.html` file that we show from the Settings screen. Sorry for the confusion!
